modules/standard/Memory/Diagnostics.chpl

```diff
@@ -0,0 +1,251 @@
+  they only modify how it is done.
+
+
+  ``memThreshold``: `uint`:
+    If this is set to a value greater than 0 (zero), only allocation
+    requests larger than this are tracked and/or reported.
+
+  ``memLog``: `string`:
+    Memory reporting is written to this file.  By default it is the
+    ``stdout`` associated with the process (not the Chapel channel
+    with the same name).  Setting this config variable to a file path
+    causes the reporting to be written to that named file instead.
+    In multilocale executions each top-level locale produces output
+    to its own file, with a dot ('.') and the locale ID appended to
+    this path.
+ */
+module Diagnostics {
+
+pragma "insert line file info"
+private extern proc chpl_memoryUsed(): uint(64);
+
+/*
+  The amount of memory returned by :proc:`locale.physicalMemory` can
+  be expressed either as individual bytes or as chunks of 2**10,
+  2**20, or 2**30 bytes.
+ */
+enum MemUnits {Bytes, KB, MB, GB};
+
+/*
+  How much physical memory is present on this locale?
+
+  Note that this is a physical quantity and does not take into account
+  things like virtual memory support which might allow for allocating
+  a larger amount, or system or user limits which might prevent
+  allocating so much.
+
+  *Note:* Unlike the other procedures in the :mod:`Diagnostics` module, this
+  one does not require memory tracking to be enabled.
+
+  :arg unit: Units in which the returned value is to be expressed.
+  :type unit: :type:`~Diagnostics.MemUnits`
+  :arg retType: Type of the returned value.  Defaults to `int(64)`.
+  :type retType: `type`
+  :returns: Size of physical memory on the locale where the call is made.
+  :rtype: `retType`
+ */
+proc locale.physicalMemory(unit: MemUnits=MemUnits.Bytes, type retType=int(64)) {
+  extern proc chpl_sys_physicalMemoryBytes(): uint(64);
+
+  var bytesInLocale: uint(64);
+
+  on this do bytesInLocale = chpl_sys_physicalMemoryBytes();
+
+  var retVal: retType;
+  select (unit) {
+    when MemUnits.Bytes do retVal = bytesInLocale:retType;
+    when MemUnits.KB do retVal = (bytesInLocale:retType / 1024):retType;
+    when MemUnits.MB do retVal = (bytesInLocale:retType / (1024**2)):retType;
+    when MemUnits.GB do retVal = (bytesInLocale:retType / (1024**3)):retType;
+  }
+
+  return retVal;
+}
+
+/*
+  How much memory is this program currently using on this locale?
+
+  This is the amount of memory known to be currently allocated on the
+  calling top-level (network-connected) locale by the program, through
+  Chapel mechanisms.  It does not include memory allocated directly
+  from the system, outside of Chapel mechanisms, such as allocations
+  made by code written in other languages and linked into the program.
+
+  :returns: Amount of allocated memory, in bytes.
+  :rtype: `uint(64)`
+ */
+proc memoryUsed() {
+  return chpl_memoryUsed();
+}
+
+/*
+  Print detailed information about allocated memory to ``memLog``.
```

Comment:
Referencing an earlier note: if memLog was mocked up to be more like an actual config in the documentation, we could link to it here and in the later places that reference it